java/monkey/src/main/java/root/parser/Parser.java

```diff
@@ -94,14 +95,96 @@ private Statement parseLetStatement() throws ParserException {
         return letStatement;
     }
 
+    private ExpressionStatement parseExpressionStatement() throws ParserException {
+        ExpressionStatement statement = new ExpressionStatement(this.currentToken);
+
+        statement.setExpression(this.parseExpression(OperatorPrecedence.LOWEST));
+
+        if (this.peekTokenIs(TokenType.SEMI)) {
+            this.nextToken();
+        }
+
+        return statement;
+    }
+
+    private void noPrefixParseFnError(TokenType type) throws ParserException {
+        throw new ParserException("No prefix parse function found for " + type);
+    }
+
+    private Expression parseExpression(OperatorPrecedence precedence) throws ParserException {
+        ParserSupplier<Expression> prefixFn = prefixParseFn();
+
+        if (prefixFn == null) {
+            noPrefixParseFnError(currentToken.type());
+        }
+        Expression leftExpression = prefixFn.get();
+
+        while (!peekTokenIs(TokenType.SEMI) && precedence.compareTo(peekPrecedence()) < 0) {
+            ParserFunction<Expression, Expression> infixFn = infixParseFn(peekToken.type());
+
+            if (infixFn == null) {
+                break;
+            }
+
+            nextToken();
```

Comment:
I'm still not a fan of `nextToken()`. It feels like it should return the next token, instead of increment.
Maybe we swap to `skipNextToken(): void` and `nextToken(): Token`?

Reply:
I kinda agree, but this is the name the author gave the function, and I believe there is some value in keeping the parallels between ours and the authors implementation.

Reply:
Perhaps another name like `proceedToNextToken()` ? I feel like `skipNextToken()` gives the impression we're ignoring something

Reply:
Sorry for taking a moment to respond. Yeah that works @RafRunner . After that change I'll approve.

Reply:
@RafRunner is that a change we're okay making? Just checking around the code.

Reply:
@Crain-32 Sorry for taking a little while, Github didnd't notify me and I was a bit busy. My next PR might be the full book completion.

Reply:
@RafRunner just tacking this on before I approve. That works for me. Thanks for doing all this work on the Java side, I really appreciate it.